Tasks/NpmAuthenticate/npmauth.ts

```diff
@@ -0,0 +1,136 @@
+    let endpointIds = tl.getDelimitedInput(constants_1.NpmAuthenticateTaskInput.CustomEndpoint, ',');
+    if (endpointIds && endpointIds.length > 0) {
+        endpointRegistries = endpointIds.map(e => npmregistry_1.NpmRegistry.FromServiceEndpoint(e, true));
+    }
+    let LocalNpmRegistries = await util.getLocalNpmRegistries(workingDir);
+    
+    
+    let npmrcFile = fs.readFileSync(npmrc, 'utf8').split('\n');
+    for (let fileLine of npmrcFile) {
+        //For each registry in the .npmrc file (every line with "registry=")
+        //If it's got a service endpoint, use that. If not, use a local registry.
+        //If it's got one of those, add the credentials into the .npmrc file
+        if (fileLine.search('registry=') > -1) {  
+            let keyAndValue = fileLine.split('=');
+            let registryURL = URL.parse(keyAndValue[1]);
+            let registry;
+            if (endpointRegistries && endpointRegistries.length > 0) {
+                for (let serviceEndpoint of endpointRegistries) {
+                    let serviceURL = URL.parse(serviceEndpoint.url);
+                    if (serviceURL.hostname == registryURL.hostname) {
+                        console.log(tl.loc("AddingEndpointCredentials", registryURL.hostname));
+                        registry = serviceEndpoint;
+                        npmrcFile = clearFileOfReferences(npmrc, npmrcFile, serviceURL);
+                        break;
+                    }
+                }
+            }
+            if (!registry) {
+                for (let localRegistry of LocalNpmRegistries) {
+                    let localURL = URL.parse(localRegistry.url);
+                    if (localURL.hostname == registryURL.hostname) {
+                        console.log(tl.loc("AddingLocalCredentials"))
+                        registry = localRegistry;
+                        npmrcFile = clearFileOfReferences(npmrc, npmrcFile, localURL);
+                        break;
+                    }
+                }
+            }
+            if (registry) {
+                if (registry.authOnly === false) {
+                    tl.debug(tl.loc('UsingRegistry', registry.url));
+                    util.appendToNpmrc(npmrc, `registry=${registry.url}\n`);
+                }
+                tl.debug(tl.loc('AddingAuthRegistry', registry.url));
+                util.appendToNpmrc(npmrc, `${registry.auth}\n`);
+                npmrcFile.push(`${registry.auth}\n`);
+            }
+            else {
+                console.log(tl.loc("IgnoringRegistry", registryURL.hostname ));
+            }
+        }
+    }
+}
+
+
+main().catch(error => {
+    tl.rmRF(util.getTempPath());
+    tl.setResult(tl.TaskResult.Failed, error);
+});
+
+
+function clearFileOfReferences(npmrc, file, url) {
+    let redoneFile = file;
+    let warned = false;
+    for (let i = 0; i < redoneFile.length; i++) {
+        if (file[i].search(url.hostname) != -1 && file[i].search('registry=') == -1) {
```

Comment:
Same comment about the hostname, it would make sense to use the nerfDart logic here and just search for that.